### PR TITLE
Invalidate-refresh-token

### PR DESCRIPTION
**Pull request by elisehuard**
* New route to invalidate the refresh token.
* Test the invalidate token route
* Test the refresh token route
* Change the timestamps associated with the refresh tokens (`:exp` and `:iat`) to contain millisecond to differentiate two timestamps create very close in time.

### CODE REVIEW

File: src/kixi/heimdall/handler.clj [hunk 1]
```diff
@@ -97,14 +104,35 @@
       {:status 201 :body res}
       {:status 401 :body res})))
 
+(defn invalidate-refresh-token [session auth-conf refresh-token]
+  (if-let [unsigned (unsign-token auth-conf refresh-token)]
+    (let [user-uuid (java.util.UUID/fromString (:user-id unsigned))
+          refresh-token-data (refresh-token/find-by-user-and-issued session
+                                                                    user-uuid
+                                                                    (:iat unsigned))]
+      (do
```
Comment (acron0) [inline src/kixi/heimdall/handler.clj]: don't need `do` here

File: src/kixi/heimdall/handler.clj [hunk 2]
```diff
@@ -97,14 +104,35 @@
...
+        (refresh-token/invalidate! session (:id refresh-token-data))
+        [true {:message "Invalidated successfully"}]))
+    [false {:message "Invalid or expired refresh token provided"}]))
```
Comment (acron0) [inline src/kixi/heimdall/handler.clj]: This `[false {:message...]` and `[true {:message...]` occurs a lot - time for fn? `(success ...)` and `(fail ... )` ?

File: src/kixi/heimdall/handler.clj [hunk 3]
```diff
@@ -97,14 +104,35 @@
...
+
+(defn invalidate-refresh-token-route [req]
+  (let [refresh-token (-> req :params :refresh-token)
+        [ok? res] (invalidate-refresh-token (:cassandra-session (:components req))
+                                            (:auth-conf req)
+                                            refresh-token)]
+    (if ok?
+      {:status 201 :body res}
```
Comment (acron0) [inline src/kixi/heimdall/handler.clj]: How come a 201?

File: src/kixi/heimdall/handler.clj [hunk 4]
```diff
@@ -97,14 +104,35 @@
       {:status 201 :body res}
```
Comment (acron0) [inline src/kixi/heimdall/handler.clj]: see comment re 201

File: src/kixi/heimdall/handler.clj (hunk 2 above)
Comment (elisehuard) [inline src/kixi/heimdall/handler.clj]: there are a couple of exceptions where the success doesn't return a message. worth a think though.

File: src/kixi/heimdall/handler.clj (hunk 3 above)
Comment (elisehuard) [inline src/kixi/heimdall/handler.clj]: 202 means accepted but not completed, so that's not right either.  Maybe 200 is the way to go. correcting ...

Comment (acron0) [inline src/kixi/heimdall/handler.clj]: Yep, or 204

File: src/kixi/heimdall/handler.clj [hunk 5]
```diff
@@ -30,96 +24,29 @@
...
+(defn refresh-auth-token [req]
   (let [refresh-token (-> req :params :refresh-token)
-        [ok? res] (refresh-auth-token (:cassandra-session (:components req))
-                                      (:auth-conf req)
-                                      refresh-token)]
+        [ok? res] (service/refresh-auth-token (:cassandra-session (:components req))
+                                              (:auth-conf req)
+                                              refresh-token)]
     (if ok?
       {:status 201 :body res}
```
Comment (acron0) [inline src/kixi/heimdall/handler.clj]: Still not sure if refreshing should cause a 201 - does it actually return a new auth token?

Comment (elisehuard) [inline src/kixi/heimdall/handler.clj]: It does

**Comment by acron0** [comment]
:shipit:

**Pull request merged**